### ISSUE TITLE
Tracking issue for RFC 2091: Implicit caller location

### ISSUE THREAD

**Issue by aturon** [rust-lang/rust#47809]
This is a tracking issue for the RFC "Implicit caller location" (rust-lang/rfcs#2091).

**Steps:**

- [x] Implement the RFC (cc @rust-lang/compiler -- can anyone write up mentoring instructions?)
- [ ] Adjust documentation ([see instructions on forge][doc-guide])
- [x] Stabilization PR ([see instructions on forge][stabilization-guide])

[stabilization-guide]: https://forge.rust-lang.org/stabilization-guide.html
[doc-guide]: https://forge.rust-lang.org/stabilization-guide.html#updating-documentation

**Unresolved questions:**

* ~~If we want to support adding `#[track_caller]` to trait methods, the redirection
    pass/query/whatever should be placed after monomorphization, not before. Currently the RFC
    simply prohibits applying `#[track_caller]` to trait methods as a future-proofing measure.~~

* ~~Diverging functions should be supported.~~

* ~~The closure `foo::{{closure}}` should inherit most attributes applied to the function `foo`, in
    particular `#[inline]`, `#[cold]`, `#[naked]` and also the ABI. Currently a procedural macro
    won't see any of these, nor would there be anyway to apply these attributes to a closure.
    Therefore, `#[rustc_implicit_caller_location]` currently will reject `#[naked]` and ABI, and
    leaving `#[inline]` and `#[cold]` mean no-op. There is no semantic reason why these cannot be
    used though.~~

**Comment by eddyb** [comment]
> If we want to support adding `#[track_caller]` to trait methods,

This refers to `impl` of a `trait`, not the declarations within the `trait`, right?

IMO the easiest way to implement this involves no procedural macros or MIR passes.
We'd just pass extra arguments on direct calls, and require a shim to get function pointers of the function. That would also surely work with trait methods, since it'd be post-monomorphization.

To explain, the shim (which we may already have some variation of) would "just" do the direct call to the function, which would pass the location of the shim, the same as the original function.

**Comment by nikomatsakis** [comment]
> This refers to impl of a trait, not the declarations within the trait, right?

I can't remember. =) But I suspect so. I don't think there's anything special about "trait methods" per se -- it's really about **dynamic dispatch**.

**Comment by nikomatsakis** [comment]
@kennytm had a prototype implementation:

https://github.com/kennytm/rust/tree/caller-info-4

Maybe @kennytm you can summarize the approach you took? Do you think you'll have time to rebase etc? I'd like ideally to get @eddyb to buy in to the overall approach. =)

**Comment by kennytm** [comment]
The prototype implementation works like this (note that `#[track_caller]` was called `#[implicit_caller_location]` there):

1. First there will be an AST transformation pass (implemented as an attribute proc-macro in ` src/libsyntax_ext/implicit_caller_location.rs`) which expands:

    ```rust
    #[track_caller]
    #[...attributes...]
    fn foo(...) -> T {
        /* code... */
    }
    ```

    into
    
    ```rust
    #[rustc_track_caller]
    #[inline]
    #[...attributes...]
    fn foo(...) -> T {
        let __closure = |__location| { /* code... */ };
        FnOnce::call_once(__closure, intrinsics::caller_location())
    }
    ```

    The purpose of this pass is to conveniently make up a new `DefId`. 

2. Create a new MIR inlining pass (main implementation in `src/librustc_mir/transform/inline.rs`). This pass does two things:

    1. Force-inline any functions calls `foo()` where the attributes of the callee `foo` contains `#[rustc_track_caller]`.
    2. If `intrinsics::caller_location()` is called :-

        * If it is used from the `__closure`, replace it by the argument `__location`. This is to propagate the caller location.
        * Otherwise, replace the call with the caller's location span.



3. Define the `caller_location()` intrinsic.

    ```rust
    fn caller_location() -> core::panic::Location<'static>;
    ```

4. Because `Location` now needs to be known by the compiler, make the Location struct a lang-item.
5. Update the standard library to use track-caller features:

    1. Change `panic!` to use `caller_location()` (or its safe counterpart, `Location::caller()`)
    2. Add `#[track_caller]` to `unwrap()` and `expect()` etc.

6. Implement the `-Z location-detail` flag (search for `location_detail`) so that the filename/line/column can be redacted.

**Comment by eddyb** [comment]
I think a syntactical transformations is unnecessary because we can instead change the "direct call ABI" (vs reifying to a fn pointer, which includes trait vtables via miri, which could go through a MIR shim). MIR inlining should *not* be needed.

**Comment by kennytm** [comment]
@eddyb So you are suggesting to change the `extern "Rust"` ABI to always pass an extra argument?

**Comment by kennytm** [comment]
@eddyb Maybe you could write down some mentoring instructions i.e. which files to look at 😊

**Comment by nikomatsakis** [comment]
@eddyb do it! do it! I want this feature.

**Comment by Centril** [comment]
@eddyb How are the mentoring instructions coming along?

**Comment by eddyb** [comment]
Oh, I don't recall exactly what happened here.

For making static calls do something different, `src/librustc_codegen_ssa/mir/block.rs` is the place you need to change, while for shims, `src/librustc_mir/shim.rs` is where their MIR is computed.

But the rest of the pieces, I don't know off the top of my head where they happen.

#54183 added a virtual-call-only shim, which is close to what this needs, so that can be used for inspiration, but here we need *both* reification to `fn` pointers *and* virtual calls.

We *could* start by disallowing reification/virtualization of such functions, and only implement the static dispatch ABI changes.

**Comment by arielb1** [comment]
let x = String::new();
match y {
Some(a) if ( drop(x); } => {}
None if { drop(x); } => {}
}
acd2fc4c63
good commit - acd2fc4c63

> We *could* start by disallowing reification/virtualization of such functions, and only implement the static dispatch ABI changes.

I don't expect this attribute to be usable on "real" functions without us implementing reification.

But for implementing this, I think one good strategy would be:

## Not supporting trait methods

For now, I don't think there is sufficient reason to support implicit caller location on trait methods, and supporting that is fairly complicated. So as @aturon said on the head PR, just don't do it.

## Step 0: dealing with the attribute.

You should make sure that the `#[blame_caller]` attribute "works". This means that:

1. It emits a feature-gate error unless the correct feature-gate is enabled (see [the feature guide] for more information about feature gates).
2. It emits an error when used with odd syntax (e.g. `#[blame_caller(42)]`),
3. It emits an error when used on something that is not a free fn/inherent impl fn.
4. It does not emit an error when used "correctly'.
5. As the RFC says: using `#[naked]` or `extern "ABI"` together with `#[rustc_implicit_caller_location]` should raise an error.

I'm not particularly sure what's the exact way to do this, but you can maybe find some PR that implemented an attribute for that, or ask on Discord.

[the feature guide]: https://forge.rust-lang.org/feature-guide.html

Make sure to add tests.

## Step 1: have reified methods and direct calls go to different LLVM functions

At this stage, I won't even add a parameter, just have the call-sites go through different paths.

So the virtual-call-only shim at #54183 is a good model for how things need to be done. You'll need to add a `ReifyShim` that is similar to `VirtualShim` in behavior (see that PR). You'll probably need to split [`ty::Instance::resolve`] to `ty::Instance::resolve_direct` and `ty::Instance::resolve_indirect` (in addition to the already-existing `ty::Instance::resolve_vtable`) and adjust the call-sites. Have `resolve_indirect` return the shim if it is needed (i.e., when calling a `#[blame_caller]` function).

[`ty::Instance::resolve`]: https://github.com/rust-lang/rust/blob/8660eba2b9bec5b0fe971b7281f79e79c2df2fae/src/librustc/ty/instance.rs#L246-L296

## Step 2.0: add a location parameter to the MIR of `#[blame_caller]`

MIR construction is in [rustc_mir::build]. There's already code there handling implicit arguments, which you could work with.

You'll need to fix [`ty::Instance::fn_sig`] to also contain the new type.

[rustc_mir::build]: https://github.com/rust-lang/rust/blob/8660eba2b9bec5b0fe971b7281f79e79c2df2fae/src/librustc_mir/build/mod.rs#L97
[`ty::Instance::fn_sig`]: https://github.com/rust-lang/rust/blob/8660eba2b9bec5b0fe971b7281f79e79c2df2fae/src/librustc/ty/instance.rs#L110-L123
    
## Step 2.1: add an "empty" location parameter to calls from the shim.

Change the shim you added in Step 1 to pass an "empty" location parameter - some "dummy" that represents a call that uses a shim. See the other shims for how to generate things in shims. Ask @eddyb if you are fighting const eval.

## Step 2.2: add a value to the location parameter to direct calls.

Find all the calls to `ty::Instance::resolve_direct` you added previously, and make them pass the "correct" location parameter to the function. These callsites can poobably already modify parameters because of `InstanceDef::Virtual`, you should be able to just follow that.

Make sure that when `#[track_caller]` calls are nested, you pass the location of your call-site, rather than *your* location. Add a test for that. Also add a test that when one of the calls in the middle is not `#[track_caller]`, the call-site is not passed.

You should probably add a function on `ty::Instance` that tells you whether you need to add the caller location.

## Step 3: wire up `intrinsic::caller_location`

See some other intrinsic for how to type-check it. Make sure it can't be called from functions that aren't `#[trace_caller]`!

I think it might even be the best to do the wiring in MIR construction, like the [`move_val_init`] intrinsic - that way you wouldn't even have to convince MIR optimizations not to dead-code-eliminate your new parameter or something evil like that, and to make it not be affected by "plain" MIR inlining you'll only have to change the `ty::Instance::resolve_direct` callsite there (you *did* make MIR inlining use `resolve_direct`, did you?).

[`move_val_init`]: https://github.com/rust-lang/rust/blob/8660eba2b9bec5b0fe971b7281f79e79c2df2fae/src/librustc_mir/build/expr/into.rs#L293-L320

## Step 4: You have it!

Now you only need to implement the library support.

**Comment by flavius** [comment]
Is anyone working on this?

**Comment by ayosec** [comment]
If nobody is working on this I would like to try to implement it.

Do you think that it is feasible as a first contribution to the compiler?

Are the steps in https://github.com/rust-lang/rust/issues/47809#issuecomment-443538059 still valid?

**Comment by cramertj** [comment]
Yes, those steps still sound reasonable. This would be a difficult first contribution, but if you ask questions on https://rust-lang.zulipchat.com/# and ping @eddyb and @arielb1 I'm sure you'd be able to start making progress, and there are likely others who could help out.

**Comment by ayosec** [comment]
I'm sorry for the delay. I couldn't have time to work on this until now =\

I have completed the step 0 described in https://github.com/rust-lang/rust/issues/47809#issuecomment-443538059 . The progress is available in [my branch](https://github.com/rust-lang/rust/compare/master...ayosec:rfc2091), though I guess that I will have to wait until the pull-request is open to know is everything is OK.

About error codes: if I understand correctly, every error emitted by the compiler has its unique code, so I have to register a new `E0###` for every error. I assume that the final code is assigned only when patch is ready to be merged into master, so I'm using dummy codes (`E0900`, `E0901`, ...) for the new errors.

**Comment by anp** [comment]
I'm interested in helping with this, are you still working on it @ayosec? If so, is there any way for me to contribute?

**Comment by ayosec** [comment]
> are you still working on it @ayosec?

Yes. I'm sorry for the lack of updates. During August I had almost no time, and I'm trying to start to continue now. At this moment I'm working with the shim to add the location parameter.

> If so, is there any way for me to contribute?

Right now I spend most of the time reading and understanding how everything works, so there is not so much code to write.

Something that would be very helpful is writing the documentation for the new errors (like [this](https://github.com/rust-lang/rust/compare/master...ayosec:rfc2091#diff-e6786747070ada32f620caf54905df92) or [this](https://github.com/rust-lang/rust/compare/master...ayosec:rfc2091#diff-0705e89a698af91d9dff388d648c5d00), at the moment), since I'm not very fluent in English.

**Comment by oli-obk** [comment]
Ping @ayosec do you have time to address this or is it OK with you if someone else takes over?

**Comment by ayosec** [comment]
I'm still working on it very slowly, but feel free to take it if you want. I can try to implement another issue once I get more time.

**Comment by anp** [comment]
I'm eager to pick this up, opened a [PR starting from your branch](https://github.com/rust-lang/rust/pull/65037).

@Centril recommended splitting this into a few PRs, so here are the remaining TODOs I summarized from @arielb1's instructions:

## add separate LLVM path

* [ ] add a `ReifyShim` that is similar to `VirtualShim` in behavior (see #54183)
* [ ] split `ty::Instance::resolve` to `ty::Instance::resolve_direct` and `ty::Instance::resolve_indirect` (leave `ty::Instance::resolve_vtable` alone)
* [ ] `resolve_indirect` returns the shim if calling a `#[track_caller]` function

## add location param

* [ ] add a location parameter to MIR of `#[track_caller]` in rustc_mir::build (reference existing implicit arguments work if needed)
* [ ] fix `ty::Instance::fn_sig` to contain the new type
* [ ] add a "dummy" location parameter to calls from `ReifyShim` (see other shims for generating things in shims, ask @eddyb if fighting consteval)
* [ ] pass the "correct" location parameter to the function in all calls to `ty::Instance::resolve_direct`
* [ ] test that directly nested `#[track_caller]`s pass the same callsite
* [ ] test that indirectly nested `#[track_caller]`s pass distinct callsites
* [ ] add method on `ty::Instance` to check whether caller location is needed

##  intrinsic::caller_location

* [ ] wire up the intrinsic in MIR construction (reference move_val_init intrinsic)
* [ ] test it can't be called from non-caller-tracked functions (reference other intrinsics for how to type-check)
* [ ] make sure MIR inlining uses resolve_direct, "change the ty::Instance::resolve_direct callsite there"

## library support

* [ ] figure out how to do this

**Comment by eddyb** [comment]
> add a location parameter to MIR of `#[track_caller]` in `rustc_mir::build` (reference existing implicit arguments work if needed)

A good example of this that landed recently is the `VaList` you get for something like this:
```rust
extern "C" fn foo(x: T, args: ...) {}
```

There it's `c_variadic` in the signature, instead of an attribute, but it's similar in that HIR and MIR bodies have one more input than the signature and a bunch of places have to account for it (see #63492).

**Comment by anp** [comment]
## library support

* [ ] safe intrinsic wrapper `core::Location::caller()`
* [ ] annotate:
  * [ ] Option::expect
  * [ ] Option::unwrap
  * [ ] Result::expect_err
  * [ ] Result::expect
  * [ ] Result::unwrap_err
  * [ ] Result::unwrap
* [ ] anything else?

## questions

Are there any stability concerns for using `#[track_caller]` in the standard library?

The RFC mentions several things related to the MIR inlining approach described there. These include limiting the number of `std` functions annotated because of inlining-based binary bloat, and disallowing taking caller-tracked fn pointers and marking trait impls as tracked because of lack of a post-monomorphization inlining pass.

However, based on @arielb1's instructions above we're not currently taking that implementation approach and IIUC the implementation won't have those inherent limitations. Some potential items for a follow-up RFC(s):

* revising RFC 2091's reference explanation to either document the implicit-argument-and-reification-shim approach we're taking or to refer to some compiler docs on it
* allowing `#[track_caller]` on trait methods, detailed explanation of point-of-reification caller location under dynamic dispatch
* changing the type signature of `core::panicking::panic_fmt` to accept a `&'static core::panic::Location` instead of the `&str,u32,u32` (this was suggested in 2091, it's a perma-unstable API as far as I can tell)
* expanding the list of annotated `std` functions

**Comment by cramertj** [comment]
> If I understand stability in std correctly, we can't migrate panic! to use Location::caller() until the latter has stabilized. Is this correct?

No, it's possible to use unstable functionality from inside std (e.g. inside the `panic!` macro).

**Comment by anp** [comment]
Is it different for macro expansions? I would naively assume so.

**Comment by cramertj** [comment]
No-- let's follow up on zulip.

**Comment by anp** [comment]
Awesome, thanks for clarifying! I edited the library support todo list appropriately.

**Comment by anp** [comment]
I have [a PR](https://github.com/rust-lang/rust/pull/65664) ready for review implementing `core::intrinsics::caller_location` and preparing the panicking infrastructure to use its return value. This lays the groundwork for propagating caller location as a single pointer-sized implicit variable through `#[track_caller]` functions, which is up next.

I'm using this [paper doc](https://paper.dropbox.com/doc/track_rfc_2091_impl-notes--Am~iTmS_9UmQVUK63kwMYAQDAQ-rwCdRc2fi2yvRZ7syGZ9q) for tracking.

**Comment by anp** [comment]
The PR I linked above has been approved and is towards the top of the homu queue (fingers crossed!). I've opened [a PR](https://github.com/rust-lang/rust/pull/65881) following that up with the start of a real implementation for the `#[track_caller]` attribute. Right now it only includes a functional const implementation and failing tests for a [not-yet-extant codegen implementation](https://paper.dropbox.com/doc/track_rfc_2091_impl-notes--Anf1NwnIb0xcRv31YLIadyj0Ag-rwCdRc2fi2yvRZ7syGZ9q#:uid=863513134494965680023183&h2=TODO-actually-pass-location-to).

**Comment by comex** [comment]
Awesome work.

Could we also add a method like `Location::current()` that always returns the location of the code that called `Location::current()`, even if it happens to be within a `#[track_caller]` function?

That way, I think e.g. `line!()` could be replaced with a real macro that expands to `Location::current().line()`.

Admittedly, although it sounds straightforward, it seems like it might be awkward to implement, given the way we wrap intrinsics in wrapper functions.

**Comment by eddyb** [comment]
It's probably far too heavy for `line!()`, but I think `(|| Location::caller())()` would work even in a `#[track_caller]` function.

**Comment by anp** [comment]
The PR [implementing the attribute](https://github.com/rust-lang/rust/pull/65881) is on its way to land. I'm working on a follow-up to lay the groundwork for annotating functions in std, hopefully I can land that before the next beta (and thus bootstrap) branch cut.

**Comment by anp** [comment]
I have [another PR](https://github.com/rust-lang/rust/pull/67137) open which I think will be the last one before we can start annotating the standard library. Ideally we can get it in before the next beta branch cut, as that will also allow us to un-cfg all of the existing non-bootstrap-compatible code in core and std.

**Comment by eddyb** [comment]
I don't think we need to get anything into beta other than the compiler changes (which all landed other than that miri change you mentioned in a comment).

That is, once beta is promoted, anything that is `cfg(not(bootstrap))` on beta can/needs to become the only variant on master, because `cfg(bootstrap)` on master is a fully bootstrapped and released beta.

**Comment by nikomatsakis** [comment]
Nominating for lang team meeting to update about the awesome progress here, that's all.

**Comment by anp** [comment]
The [PR to have `panic!` use the new intrinsic always](https://github.com/rust-lang/rust/pull/67137) has landed! I'm working on a branch now to annotate these functions in std:

* Option::expect
* Option::unwrap
* Result::expect_err
* Result::expect
* Result::unwrap_err
* Result::unwrap

The RFC also listed several implementations of `Index` and `IndexMut`, but it wasn't approved because of implementation limitations:

> This RFC only advocates adding the #[track_caller] attribute to the unwrap and expect functions. The index and index_mut functions should also have it if possible, but this is currently postponed as **it is not investigated yet how to insert the transformation after monomorphization**.

Thing is, the current implementation is significantly different from the approved RFC's proposal and actually is compatible with trait methods to my knowledge (it is after monomorphization IIUC).

I'm interested in adding the attribute to the indexing functions, should I submit an "amended RFC" describing the new implementation and seek consensus on applying the attribute to trait methods? I've already thought about doing that since the implementation has deviated so much from the original proposal.

**Comment by anp** [comment]
Have that PR open with the panic messages working!!!

https://github.com/rust-lang/rust/pull/67887

**Comment by lzutao** [comment]
About the [code bloat][1] drawback in the RFC, would setting `RUST_BACKTRACE=1` make it 
unnoticeable?

[1]: https://rust-lang.github.io/rfcs/2091-inline-semantic.html#code-bloat

**Comment by anp** [comment]
> About the [code bloat](https://rust-lang.github.io/rfcs/2091-inline-semantic.html#code-bloat) drawback in the RFC, would setting `RUST_BACKTRACE=1` make it
> unnoticeable?

I don't think so, because `RUST_BACKTRACE` has effect at runtime, not compile time IIUC.

**Comment by anp** [comment]
Alright, https://github.com/rust-lang/rust/pull/67887 has landed! Going to manually verify everything is working with tomorrow's nightly.

@nikomatsakis what do you think should be the process to form a plan for `#[track_caller]` in trait impls? Mentioned [above)(https://github.com/rust-lang/rust/issues/47809#issuecomment-570825317).

**Comment by nikomatsakis** [comment]
@rustbot modify labels to +I-Nominated

Nominating for lang-team discussion on @anp's final question. Does anyone object to extending `#[track_caller]` to work in trait impls. Apparently the new discussion strategy (which I am still investigating) can handle this just fine.

My opinion: we should definitely do it! But I would like to see the new impl strategy written up for the rustc-guide, as I'd like to know how it works. (I'm not sure how much of that belongs in the reference, too? At least some of it.)

**Comment by anp** [comment]
Re: a write-up, I planned to write about the attribute for the reference and to write some info about the implicit arguments for the book. I had considered these important for stabilization and am going to do them regardless. If that's all that's required to finish the last bit of proposed work from the RFC, I'm excited!

**Comment by nikomatsakis** [comment]
@anp 

> write some info about the implicit arguments for the book.

what is "the book" in that sentence?

One thing I would really encourage is that you document the high-level picture of the impl for rustc-guide.

**Comment by nikomatsakis** [comment]
We discussed this in our [lang team meeting today](https://github.com/rust-lang/lang-team/blob/master/minutes/2020-01-09.md). We approve of extending this to trait items. This was based on an understanding that it falls out fairly natural from the implementation -- I gave a summary of what I understand based on skimming the PRs plus earlier conversations with @eddyb.

**Comment by anp** [comment]
Ah yes, i meant rustc-guide when I said ‘the book’ - I’ve been referring to it much more than TRPL and forgot myself 🤣.

**Comment by eddyb** [comment]
This example has regressed (including on stable by now): ([playground](
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d5d1dd7fc4b97a199f8501d9cbabcfa0))
```rust
fn main() {
    std::collections::VecDeque::<String>::with_capacity(!0);
}
```
The location reported is now `<::core::macros::panic macros>:3:10`.

In general, we don't track enough information to recover the invocation site, when we have to codegen the MIR cross-crate. @Aaron1011 made me realize this is a problem in https://github.com/rust-lang/rust/pull/68965#issuecomment-583781140.

A cheap hack we could do is replace the macro span with the invocation one, when serializing `Span`s inside MIR (or `Span`s in general?).

cc @petrochenkov

**Comment by anp** [comment]
Status update:

1. the bug @eddyb mentioned in https://github.com/rust-lang/rust/issues/47809#issuecomment-583784847 has a proposed fix in https://github.com/rust-lang/rust/pull/68965. I don't think there's any help I can offer, but I'm available if there is.
2. I've opened a [tentative PR to the reference](https://github.com/rust-lang/reference/pull/742) so that docs are ready and reviewed whenever stabilization happens.
3. I've opened a [draft PR to rustc-guide](https://github.com/rust-lang/rustc-guide/pull/582). I have some more handwritten notes to add to it and will ask for review when it's fleshed out.
4. I've opened a [PR to allow the attribute in traits and adding it to indexing traits in std](https://github.com/rust-lang/rust/pull/69251).
5. I believe once the trait impl lands that all [unresolved questions](https://github.com/rust-lang/rfcs/blob/master/text/2091-inline-semantic.md#unresolved-questions) from the RFC will be resolved. Woo!

I'll be keeping those PRs warm and adding to the rustc-guide section for now. Modulo new bugs, I'm hoping that we'll be able to let this bake for a couple of releases and then the next active task will be a stabilization report. Am I missing anything?

**Comment by eddyb** [comment]
> the bug @eddyb mentioned in https://github.com/rust-lang/rust/issues/47809#issuecomment-583784847 has a proposed fix in #68965

No, it was brought up on #68965, but the fix is to track hygiene info cross-crate, which last I heard @Aaron1011 was working on.

**Comment by Aaron1011** [comment]
I have some code locally which serializes hygiene information, but it's currently blocked on https://github.com/rust-lang/rust/pull/68941

**Comment by petrochenkov** [comment]
>https://github.com/rust-lang/rust/issues/47809#issuecomment-583784847

[triagebot] I don't know anything about the treatment of spans in MIR inlining.

**Comment by eddyb** [comment]
@petrochenkov There is 0 MIR inlining going on in my example, sorry for the confusion (@Aaron1011 brought up in regards to MIR inlining but the problem is always codegen of a function defined in another crate).
MIR inlining isn't even on by default yet.

The problem I noted is entirely because we can't get the invocation span (because the `ExpnId` is gone cross-crate), so the `panic::Location` refers to inside the macro instead of the invocation site of the macro.

**Comment by est31** [comment]
cc #69977

**Comment by anp** [comment]
Checking in and it looks like #68941 has merged, @Aaron1011 is there a PR I can follow for the MIR-serialized-hygiene-info?

**Comment by Aaron1011** [comment]
@anp: The next step is https://github.com/rust-lang/rust/pull/70077, which is currently in the merge queue. I plan to open an WIP PR for hygiene serialization fairly soon.

**Comment by anp** [comment]
A question came up in code review re: the attribute's behavior in traits with specialization, I filed [an issue for follow-up](https://github.com/rust-lang/rust/issues/70293) since specialization is unstable and no currently planned uses of the attribute require an answer.

**Comment by anp** [comment]
https://github.com/rust-lang/rust/pull/69251 is in the merge queue, implementing the attribute in traits.

**Comment by anp** [comment]
Support for the attribute in traits has landed, currently working on landing Index(Mut) support in #70234.

**Comment by anp** [comment]
I finished the first full draft of the [section of the rustc-dev-guide](https://github.com/rust-lang/rustc-dev-guide/pull/582) documenting the feature, the PR is still open. I'm sure there are aspects of this that aren't covered sufficiently or additional material that could be helpful. Feedback on the PR very much appreciated!

I *think* that the only remaining implementation question is how to handle `-Zlocation-detail-control={file,line,column}` [from the RFC](https://github.com/rust-lang/rfcs/blob/master/text/2091-inline-semantic.md#location-detail-control). I'm not sure how useful the multiple levels of granularity are.

It seems like there are two motivations that have come up for the option:

1. to preserve pre-existing optimization opportunities in cold error-handling codepaths
2. to mitigate binary size bloat from additional file paths being included in the binary (which aren't stripped w/ debug info)

(1) was a concern at the time of the RFC but we've not yet seen any performance regressions to suggest that this is happening to the serious detriment of any projects (projects that are very sensitive to code size of panicking already avoid a lot of that code at all costs).

(2) seems like a legitimate concern and could be addressed with a simpler `-Zredact-caller-location` flag to make all `Location`s the same. Incidentally this might allow for even more opportunities in (1).

IIUC in order to implement this we need to preserve the types/layouts/etc of all of the APIs since e.g. std is precompiled for most builds and the binaries must be compatible. Is it as simple as changing the allocation functions to replace filenames with `<redacted>` and setting the line/column numbers to 0?

**Comment by eddyb** [comment]
I was suggesting elsewhere that we could be using sections. Specifically, two:
* the string data for filenames would go into one section, which could be discarded
* the `Location`s i.e. `(&str, u32, u32)` would go into another section, which would be zeroed
  * `Location` will have to use `Option<&str>` for filenames, instead of `&str`, for this to be sound

However, I'm not sure there's a good way to automate that treatment, and it'd be a bit excessive for us to make our tools for this. But we do control the last linker invocation, so we could forward to it?

> Is it as simple as changing the allocation functions to replace filenames with <redacted> and setting the line/column numbers to 0?

For `RUSTFLAGS=-Zredact-caller-location` that can't hide `Location`s already compiled into `core`/`alloc`/`std`, sure, and I suppose for many purposes that might be enough.

Should we have a single flag that also affects `file!()`/`line!()`/`column!()` too? What about debug info?

**Comment by Centril** [comment]
> (2) seems like a legitimate concern and could be addressed with a simpler `-Zredact-caller-location` flag to make all `Location`s the same. Incidentally this might allow for even more opportunities in (1).

This seems like a reasonable compromise. 👍

**Comment by anp** [comment]
[Update to the binary size investigation](https://github.com/rust-lang/rust/issues/70579#issuecomment-609475889): a custom section shows *some* measurable size usage but I'm unsure whether it's enough to gate stabilization on a mitigation strategy. Input on the measurement issue appreciated!

**Comment by anp** [comment]
Going over my notes, and the binary size was the last TODO for this stabilization report.

# Stabilization Report

## Summary

From the [rustc-dev-guide chapter][dev-guide]:

> Approved in [RFC 2091], this feature enables the accurate reporting of caller location during 
> panics initiated from functions like `Option::unwrap`, `Result::expect`, and `Index::index`. This
> feature adds the `#[track_caller]` attribute for functions, the `caller_location` intrinsic, and 
> the stabilization-friendly `core::panic::Location::caller` wrapper.

The attribute adds an implicit caller location argument to the ABI of annotated functions, but does
not affect the type or MIR of the function. We implement the feature entirely in codegen and in the
const evaluator.

## Tests

All of the tests for this feature live under [src/test/ui/rfc-2091-track-caller][tests] in the repo.

Noteworthy cases:

* [use of attr in std]
  * validates user-facing benefit of the feature
* [trait attribute inheritance]
  * covers subtle behavior designed during implementation and not RFC'd
* [const/codegen equivalence]
  * this was the result of a suspected edge case and investigation
* [diverging function support]
  * covers an unresolved question from the RFC
* [fn pointers and shims]
  * covers important potential sources of unsoundness

## Documentation

The rustc-dev-guide now has a chapter on [Implicit Caller Location][dev-guide].

I have an [open PR to the reference][attr-reference-pr] documenting the attribute.

The intrinsic's [wrapper] includes some examples as well.

## Unresolved Questions

### From the RFC

> Currently the RFC simply prohibit applying #[track_caller] to trait methods as a future-proofing
> measure.

**Resolved.** See the dev-guide documentation and the tests section above.

> Diverging functions should be supported.

**Resolved.** See the tests section above.

> The closure foo::{{closure}} should inherit most attributes applied to the function foo, ...

**Resolved.** This unknown was related to specifics of the implementation which were made irrelevant
by the final implementation.

### Binary Size

I [instrumented track_caller to use custom sections][measure-size] in a local build and discovered
relatively minor binary size usage for the feature overall. I'm leaving the issue open to discuss
whether we want to upstream custom section support.

There's an [open issue to discuss mitigation strategies][mitigate-size]. Some decisions remain about
the "right" strategies to reduce size without overly constraining the compiler implementation. I'd
be excited to see someone carry that work forward but my opinion is that we shouldn't block
stabilization on implementing compiler flags for redaction.

### Specialization

There's an [open issue][specialization] on the semantics of the attribute in specialization chains.
I'm inclined to move forward with stabilization without an exact resolution here given that
specialization is itself unstable, but I also think it should be an easy question to resolve if
others disagree.

### Maturity

My final unresolved question: how long should we let the use of this feature in std be visible to
users on stable before we consider it "baked"? While users have had unwrap/expect support on stable
since 1.42, trait support and its use in std still won't be visible to stable users until 1.43
(or 1.44?). In my ideal world we'd have a stable release with this feature by the end of the summer,
but I'm not sure if that's realistic.

[dev-guide]: https://rustc-dev-guide.rust-lang.org/codegen/implicit-caller-location.html
[specialization]: https://github.com/rust-lang/rust/issues/70293
[measure-size]: https://github.com/rust-lang/rust/issues/70579
[mitigate-size]: https://github.com/rust-lang/rust/issues/70580
[attr-reference-pr]: https://github.com/rust-lang/reference/pull/742
[wrapper]: https://doc.rust-lang.org/nightly/core/panic/struct.Location.html#method.caller
[tests]: https://github.com/rust-lang/rust/tree/master/src/test/ui/rfc-2091-track-caller
[const/codegen equivalence]: https://github.com/rust-lang/rust/blob/master/src/test/ui/rfc-2091-track-caller/caller-location-fnptr-rt-ctfe-equiv.rs
[diverging function support]: https://github.com/rust-lang/rust/blob/master/src/test/ui/rfc-2091-track-caller/diverging-caller-location.rs
[use of attr in std]: https://github.com/rust-lang/rust/blob/master/src/test/ui/rfc-2091-track-caller/std-panic-locations.rs
[fn pointers and shims]: https://github.com/rust-lang/rust/blob/master/src/test/ui/rfc-2091-track-caller/tracked-fn-ptr-with-arg.rs
[trait attribute inheritance]: https://github.com/rust-lang/rust/blob/master/src/test/ui/rfc-2091-track-caller/tracked-trait-impls.rs

**Comment by scottjmaddox** [comment]
Is there a reason `file()`, `column()`, and `line()` aren't `const fn`s?

Also, is there a reason for not having `module_path`?

**Comment by nikomatsakis** [comment]
Bug that @eddyb raised relating to panics and cross-crate macro spans:

https://github.com/rust-lang/rust/issues/70963

**Comment by anp** [comment]
I've updated the stabilization report with a couple more bug links, a better opening, and a history of PRs for the feature.

**Comment by nikomatsakis** [comment]
@anp one thing I feel like I am missing is what the "scope" of this stabilization report covers. I'd like a kind of checklist of the new things that are possible now if we opt to stabilize.

I believe it is:

* `track_caller` can be placed on functions in all locations, and it will influence the results of the `Location` APIs and the output from panics and the like

but I am not sure whether the `Location` APIs themselves are, for example, covered under this stabilization. 

I think what we would normally do, also, is to create an issue dedicated to the stabilization, or perhaps a PR that actually makes the changes. That's a bit cleaner than using the tracking issue.

**Comment by anp** [comment]
Good idea! I’ll work on that.

**Comment by joshtriplett** [comment]
I'm going to temporarily un-nominate this given the request from @nikomatsakis, but please feel free to re-nominate this once updated with that information.

**Comment by ggriffiniii** [comment]
I scanned the history but didn't see this question mentioned. Has there been any discussion about moving the Location type out of the std::panic module, or even just re-exporting under a different name if it needs to stay there for backwards compatibility?

I'm really looking forward to track_caller, but almost none of my use cases are panic related and panic is a pretty eye-catching word in code that causes me to do a double-take every time I encounter it. It seems like a superficial thing to discuss compared with all the implementation details, but I just wanted to provide my experience in case that aspect has been overlooked.

**Comment by anp** [comment]
This has come to mind for me a few times and I’m glad you brought it up. Seems worth filing an issue for discussion?

**Comment by est31** [comment]
What would be non-panic-related use cases of this feature? It *only* affects panic information.

As for moving the location out of the `std::panic` module, there are three ways the standard library crates (including proc_macro) display location information. First is through `std::panic::Location`, second is `std::backtrace::Backtrace`, third is `proc_macro::Span`. They use three distinct methods of passing the data specific to their way of functioning. If `Location` is moved out of std::panic, care should be taken to avoid confusion with the other methods.

**Comment by jethrogb** [comment]
> It *only* affects panic information.

Not anymore, you can just call `Location::caller()` anywhere and use it, independently of panicking/unwinding.

**Comment by dtolnay** [comment]
Labeling <kbd>T-libs</kbd> because `std::panic::Location::caller` is covered by this tracking issue.

**Comment by est31** [comment]
@jethrogb good point, thanks.

**Comment by anp** [comment]
Alright! I've revised the PR to the reference and have opened a PR to stabilize the attribute and wrapper: https://github.com/rust-lang/rust/pull/72445.

**Comment by i-schuetz** [comment]
Sorry, what is the name of the flag? `location-detail` and `redact-caller-location` are returning "unknown flag"

**Comment by anp** [comment]
The flag needs implementation still: https://github.com/rust-lang/rust/issues/70580.

**Comment by anp** [comment]
The [stabilization PR](https://github.com/rust-lang/rust/pull/72445) just landed! The [PR to the reference](https://github.com/rust-lang/reference/pull/742) should be ready to go as well, after which I think we can close this tracking issue (at least those are the only checkboxes remaining in the top message).

**Comment by nikomatsakis** [comment]
Amazing work, @anp!

**Comment by elichai** [comment]
This is awesome!
One problem I have is that it's not possible to put `#[track_caller]` on closures, this is really useful when `thread_local` is involved because then there's a lot of logic in closures that you might want to propagate
ie
```rust
thread_local! {static DONE: bool = false;}

#[track_caller]
fn assert_done() {
    DONE.with(
        #[track_caller]
        |b| assert!(b),
    );
}

fn main() {
    assert_done();
}
```

**Comment by anp** [comment]
The restriction on closures were in the original RFC, I think as a result of the implementation proposed then. Right now, I can't think of a reason the current implementation couldn't support this but I could easily be missing something. I opened https://github.com/rust-lang/rust/issues/74042 to discuss/track.

**Comment by xd009642** [comment]
Apologies if this has been raised before, but I've been playing around with trying to track where allocations happen with something like so:

```Rust
use libc_print::libc_println;
use std::alloc::{GlobalAlloc, Layout};
use std::panic::Location;

pub struct TracedAlloc<T: GlobalAlloc> {
    pub allocator: T,
}

unsafe impl<T> GlobalAlloc for TracedAlloc<T>
where
    T: GlobalAlloc,
{
    #[track_caller]
    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
        libc_println!("Alloc {:?} at {:?}", layout, Location::caller());
        self.allocator.alloc(layout)
    }

    #[track_caller]
    unsafe fn dealloc(&self, ptr: *mut u8, layout: Layout) {
        libc_println!("Dealloc {:?} at {:?}", layout, Location::caller());
        self.allocator.dealloc(ptr, layout)
    }
}
```

However the caller location is always the line I've put `#[global_allocator]` which makes `#[track_caller]` useless in this context.

**Comment by KodrAus** [comment]
Are we happy to close this tracking issue now that #74042 has spun off work to support the attribute on closures?

**Comment by nikomatsakis** [comment]
Sure, I guess so. Thanks again @anp for seeing this through!

**Comment by schreter** [comment]
Hi. I know this is a bit off-topic and this particular issue is already closed, but I suppose, the right people are involved here :-), so someone might help.

I implemented for our project a version of `Result` which uses `Try` trait in such a way as to track the error handling locations, so we get a "call stack" of the error at the end. This works in conjunction with `#[track_caller]` very well and we see the locations the error handling took.

However, there is one major deficiency in `Location` - it only gives us the source name and line. Yes, with that, it's possible to manually look up the function where it happened, but it would be significantly simpler to evaluate bug reports by looking at the call trace with function names (we have practiced this in a large C++ project, where basically the dev support standardized on initially evaluating everything by function names call trace, ignoring line numbers). So it would come extremely handy if the `Location` could be extended with another `&str` containing the function name (ideally with generics). It can be also a mangled name, I don't care much, but the function name is important.

Before you should "backtrace!" - yes, but... We are using heavy asynchronous processing handling errors across `await`s, where the backtrace has about zero value. Similar for tracing, we can't just always trace due to performance reasons. So the error handling "call stack" is a perfect compromise - cheap and sufficiently valuable (except that it's missing the function name).

Any suggestion where/how to address this issue (`Location` extension by file name)? According to my code study of the Rust compiler code, it should basically boil down to getting the function name from the current `Span` and adding it in addition to the file name to the generated `const Location` here: https://github.com/rust-lang/rust/blob/012720ffb075a087b781325d17d1822a340a2f2a/compiler/rustc_codegen_cranelift/src/common.rs#L351

Thanks & regards,

Ivan

**Comment by jplatte** [comment]
I remember asking about this quite a while ago and getting a response along the lines of "unlikely to happen for perf or code size reasons". Unfortunately I can't find an issue about it, so I guess it must have been on Zulip.

I think creating a separate issue would be more useful than continuing this conversation here, in any case.

**Comment by schreter** [comment]
> I remember asking about this quite a while ago and getting a response along the lines of "unlikely to happen for perf or code size reasons". Unfortunately I can't find an issue about it, so I guess it must have been on Zulip.

Well, it's clear that this will increase generated constants segment by potentially quite a bit, because instead of a single string per file we'll need to store many strings per file. So it could be made optional at compilation time. But I personally don't think it'll be significantly slower (during compilation; runtime is obviously unaffected).

> I think creating a separate issue would be more useful than continuing this conversation here, in any case.

Sure. Simply a new top-level issue w/o any comments? The question is how to get it to the attention of relevant people who could do something about it.

**Comment by est31** [comment]
If you have the source code, it should be quite easy to build a tool that translates the Location to the name of the function. `Location` contains the file name, as well as line and column, so all you need to do is go to that specific line and column, and then parse the next ident. Done easily with `syn` where you can write a custom visitor with a `visit_ident` function, like [this](https://docs.rs/syn/1.0.90/syn/visit/index.html#example). Then you just compare the line/col number with the line/col number of the `Location`, done. Maybe if the source code is not public you can export the `Location` via e.g. json and import it via the tool.

**Comment by schreter** [comment]
> If you have the source code, it should be quite easy to build a tool that translates the Location to the name of the function.

Of course it could be translated. But that's another step, which makes it quite cumbersome. The compiler already knows it at compile time and aside from costing more space in the generated executable (string section), there should be no adverse effects of having the function name in the `Location`. It would in general allow building various tools which use the location for debugging purposes, which in turn would help the community in general.

Let me move this to a new issue, for further discussion.
